appinventor/components/src/com/google/appinventor/components/runtime/util/EclairUtil.java

```diff
@@ -5,30 +5,60 @@
 // http://www.apache.org/licenses/LICENSE-2.0
 package com.google.appinventor.components.runtime.util;
 
+import com.google.appinventor.components.runtime.Form;
 import com.google.appinventor.components.runtime.WebViewer;
 
 import android.app.Activity;
 import android.app.AlertDialog;
 import android.content.Context;
 import android.content.DialogInterface;
 
+import android.net.Uri;
 import android.text.InputType;
 import android.view.MotionEvent;
 import android.view.View;
 import android.webkit.GeolocationPermissions;
 import android.webkit.GeolocationPermissions.Callback;
+import android.webkit.PermissionRequest;
 import android.webkit.WebView;
 import android.webkit.WebViewClient;
 import android.webkit.WebChromeClient;
 import android.widget.EditText;
 
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
 /**
  * Helper methods for calling methods added in Eclair (2.0, API level 5)
  *
  * @author Ryan Bis
  *
  */
 public class EclairUtil {
+  /*
+   * Maps WebView resource names to more user-friendly names
+   * Used in setupWebViewPermissions
+   * TODO: Might be worthwhile to move this somewhere else.
+  */
+  private static final Map<String, String> RESOURCE_MAP;
+  private static final Map<String, String[]> PERMISSION_MAP;
+
+  static {
+    RESOURCE_MAP = new HashMap<String, String>();
+    RESOURCE_MAP.put("android.webkit.resource.AUDIO_CAPTURE", "microphone");
+    RESOURCE_MAP.put("android.webkit.resource.MIDI_SYSEX", "MIDI device");
+    RESOURCE_MAP.put("android.webkit.resource.PROTECTED_MEDIA_ID", "EME APIs");
+    RESOURCE_MAP.put("android.webkit.resource.VIDEO_CAPTURE", "camera");
+
+    PERMISSION_MAP = new HashMap<String, String[]>();
+    PERMISSION_MAP.put("location", new String[]{"android.permission.ACCESS_FINE_LOCATION",
```

Comment:
Minor code suggestion: We could reference permission name constants like Manifest.ACCESS_FINE_LOCATION, etc. by importing android.Manifest.